### ISSUE TITLE
error training asr with pretrained aalbert

### ISSUE THREAD

**Issue by ybNo1** [andi611/Self-Supervised-Speech-Pretraining-and-Representation-Learning#20]
when running pytorch-kaldi for asr training using pretrained aalbert model, something went wrong like this:
```
$ python run_exp.py cfg/AISHELL/aishell_transformer_fmllr_ft.cfg 
- Reading config file......OK!
- Chunk creation......OK!

------------------------------ Epoch 00 / 23 ------------------------------
Training train chunk = 1 / 50
Better speed can be achieved with apex installed from https://www.github.com/nvidia/apex .
[Transformer] - Pre-trained weights loaded!
[Transformer] - Number of parameters: 7182336
Traceback (most recent call last):
  File "run_exp.py", line 281, in <module>
    next_config_file,
  File "/home/bin.yang02/program/pytorch-kaldi/core.py", line 628, in run_nn
    forward_outs,
  File "/home/bin.yang02/program/pytorch-kaldi/utils.py", line 2334, in forward_model
    outs_dict[inp2] = outs_dict[inp2].view(max_len * batch_size, -1)
RuntimeError: shape '[416, -1]' is invalid for input of size 307200
```

I trained the aalbert model with the cfg file like this:
```
$ cat aalbert_aishell_fbank3L.yaml
transformer:
  input_dim: 40                                         # `int`, 39 for mfcc, 40 for fmllr, 80 for fbank, 160 for mel
  downsample_rate: 3                                    # stacked consecutive features vectors to reduce the length of input sequences by this factor.
  hidden_size: 768                                      # Size of the encoder layers and the pooler layer.
  num_hidden_layers: 3                                  # Number of hidden layers in the Transformer encoder.
  num_attention_heads: 12                               # Number of attention heads for each attention layer in the Transformer encoder.
  intermediate_size: 3072                               # The size of the "intermediate" (i.e., feed-forward) layer in the Transformer encoder.
  hidden_act: "gelu"                                    # The non-linear activation function (function or string) in the encoder and pooler. If string, "gelu", "relu" and "swish" are supported.
  hidden_dropout_prob: 0.1                              # The dropout probabilitiy for all fully connected layers in the embeddings, encoder, and pooler.
  attention_probs_dropout_prob: 0.1                     # The dropout ratio for the attention probabilities.
  initializer_range: 0.02                               # The sttdev of the truncated_normal_initializer for initializing all weight matrices.
  layer_norm_eps: "1e-12"                               # The epsilon used by LayerNorm.
  mask_proportion: 0.15                                 # mask this percentage of all spectrogram frames in each sequence at random during MAM training                        
  mask_consecutive_min: 7                               # mask this amount of consecutive frames
  mask_consecutive_max: 7                               # mask this amount of consecutive frames
  mask_allow_overlap: True                              # allow overlap masking
  mask_bucket_ratio: 1.2
  mask_bucket_ratio: 1.2                                # only used when overlap is not allowed. sample a mask from each bucket in size of [sampled mask_consecutive * mask_bucket_ratio]
  mask_frequency: 16                                    # mask maximum this amount of frequency bands, set to 0 for no frequency mask
  noise_proportion: 0.15                                # for this percentage of the time, Gaussian noise will be applied on all frames during MAM training, set to 0 for no noise
  prune_headids: None                                   # Usage: 0,1,2,12-15 will prune headids [0,1,2,12,13,14]. headids = layerid * head_num + headid_in_layer
  share_layer: True                                     # Share layer weights
  max_input_length: 0                                   # maximum input length (0 for no restriction)

optimizer: 
  learning_rate: "2e-4"                                 # Learning rate for opt. "4e-4" for 'data/libri_mel160_subword5000', "2e-4" for 'data/libri_fmllr_cmvn'
  loss_scale: 0                                         # Loss scale to improve fp16 numeric stability. Only used when apex is set to True. 0: dynamic loss scaling. positive power of 2: static loss scaling.
  warmup_proportion: 0.07                               # Proportion of training to perform linear rate warmup.
  gradient_accumulation_steps: 3                        # Number of updates steps to accumulate before performing a backward/update pass
  gradient_clipping: 3.0                                # Maximum gradient norm

dataloader:
  n_jobs: 12                                            # Subprocess used for torch Dataloader
  batch_size: 12                                        # training batch size, 12 for pre-train, 6 for cpc exp
  dev_batch_size: 12                                    # used for dev/test splits
  max_timestep: 1500                                    # Max length for audio feature (0 for no restriction), 1500 for pre-train, 3000 for downstream tasks
  
  # LIBRISEECH SETTINGS
  data_path: 'data/aishell_fmllr_cmvn'                    # Source data path, 'data/libri_fmllr_cmvn', or 'data/libri_mfcc_cmvn', or 'data/libri_mel160_subword5000' for different preprocessing features
  target_path: ''                                       # Target data path, not used when `duo_deature:False`. For reconstruction to a different feature type, for example set dataset to 'libri_linear1025_subword5000'.
  phone_path: 'data/cpc_phone'                          # phone boundary label data path for the phone classification task. set to 'data/libri_phone' or 'data/cpc_phone'
  train_set: ['train']                        # ['train-clean-100', 'train-clean-360', 'train-other-500'] for pre-training. ['train-clean-360'] or ['train-clean-100'] for libri phone exp or cpc phone exp, respectively.
  dev_set: ['dev']                                #
  test_set: ['test']                              #
  train_proportion: 1.0                                 # Currently only effect the `phone classification task`, use this percent of `train_set` for downstream task training to demonstrate mockingjay generality

runner:
  # Training options
  apex: False                                           # Use APEX (see https://github.com/NVIDIA/apex for more details)
  total_steps: 200000                                   # total steps for training, a step is a batch of update
  log_step: 2500                                        # log training status every this amount of training steps
  save_step: 10000                                      # save model every this amount of training steps
  duo_feature: False                                    # Use different input / output features during training
  max_keep: 2                                           # maximum number of model ckpt to keep during training 
```

and the asr train cfg file:

```
$ cat aishell_transformer_fmllr_ft.cfg
[cfg_proto]
cfg_proto=proto/global.proto
cfg_proto_chunk=proto/global_chunk.proto

[exp]
cmd=
run_nn_script=run_nn
out_folder=exp/aishell_transformer_fmllr
seed=1234
use_cuda=True
multi_gpu=False
save_gpumem=False
N_epochs_tr=24

[dataset1]
data_name=train
fea:fea_name=fmllr
    fea_lst=/home/bin.yang02/program/kaldi-scripts/aishell/s5/fmllr/train/feats.scp
    fea_opts=apply-cmvn --utt2spk=ark:/home/bin.yang02/program/kaldi-scripts/aishell/s5/fmllr/train/utt2spk  ark:/home/bin.yang02/program/kaldi-scripts/aishell/s5/fmllr/train/data/cmvn_speaker.ark ark:- ark:- | add-deltas --delta-order=0 ark:- ark:- |
    cw_left=0
    cw_right=0

    
lab:lab_name=lab_cd
    lab_folder=/home/bin.yang02/program/kaldi-scripts/aishell/s5/exp/tri5a_ali_train/
    lab_opts=ali-to-pdf 
    lab_count_file=auto
    lab_data_folder=/home/bin.yang02/program/kaldi-scripts/aishell/s5/fmllr/train/
    lab_graph=/home/bin.yang02/program/kaldi-scripts/aishell/s5/exp/tri5a/graph/

N_chunks=32
        
[dataset2]
data_name=dev
fea:fea_name=fmllr
    fea_lst=/home/bin.yang02/program/kaldi-scripts/aishell/s5/fmllr/dev/feats.scp
    fea_opts=apply-cmvn --utt2spk=ark:/home/bin.yang02/program/kaldi-scripts/aishell/s5/fmllr/dev/utt2spk  ark:/home/bin.yang02/program/kaldi-scripts/aishell/s5/fmllr/dev/data/cmvn_speaker.ark ark:- ark:- | add-deltas --delta-order=0 ark:- ark:- |
    cw_left=0
    cw_right=0


lab:lab_name=lab_cd
    lab_folder=/home/bin.yang02/program/kaldi-scripts/aishell/s5/exp/tri5a_ali_dev/
    lab_opts=ali-to-pdf 
    lab_count_file=auto
    lab_data_folder=/home/bin.yang02/program/kaldi-scripts/aishell/s5/fmllr/dev/
    lab_graph=/home/bin.yang02/program/kaldi-scripts/aishell/s5/exp/tri5a/graph/


N_chunks=32

[dataset3]
data_name=test
fea:fea_name=fmllr
    fea_lst=/home/bin.yang02/program/kaldi-scripts/aishell/s5/fmllr/test/feats.scp
    fea_opts=apply-cmvn --utt2spk=ark:/home/bin.yang02/program/kaldi-scripts/aishell/s5/fmllr/test/utt2spk  ark:/home/bin.yang02/program/kaldi-scripts/aishell/s5/fmllr/test/data/cmvn_speaker.ark ark:- ark:- | add-deltas --delta-order=0 ark:- ark:- |
    cw_left=0
    cw_right=0


lab:lab_name=lab_cd
    lab_folder=/home/bin.yang02/program/kaldi-scripts/aishell/s5/exp/tri5a_ali_test/
    lab_opts=ali-to-pdf 
    lab_count_file=auto
    lab_data_folder=/home/bin.yang02/program/kaldi-scripts/aishell/s5/fmllr/test/
    lab_graph=/home/bin.yang02/program/kaldi-scripts/aishell/s5/exp/tri5a/graph/


N_chunks=16

        
[data_use]
train_with=train
valid_with=dev
forward_with=test


[batches]
batch_size_train=16
max_seq_length_train=1000
increase_seq_length_train=True
start_seq_len_train=100
multply_factor_seq_len_train=2
batch_size_valid=8
max_seq_length_valid=1000


[architecture1]
arch_name=TRANSFORMER_AM
arch_proto=proto/Transformer.proto
arch_library=nn_transformer
arch_class=TRANSFORMER
arch_pretrain_file=none
arch_freeze=False
arch_seq_model=True

# Transformer.proto settings
ckpt_file=/home/bin.yang02/software/Self-Supervised-Speech-Pretraining-and-Representation-Learning/result/result_transformer/aalbert_fbank3L/states-200000.ckpt
load_pretrain=True
no_grad=False
dropout=default
spec_aug=True
spec_aug_prev=True
weighted_sum=False
select_layer=-1

# Optimizer Settings
arch_lr = 0.0002
arch_halving_factor = 0.5
arch_improvement_threshold = 0.001
arch_opt = rmsprop
opt_momentum = 0.0
opt_alpha = 0.95
opt_eps = 1e-8
opt_centered = False
opt_weight_decay = 0.0


[architecture2]
arch_name=MLP_layers
arch_proto=proto/MLP.proto
arch_library=neural_networks
arch_class=MLP
arch_pretrain_file=none
arch_freeze=False
arch_seq_model=False
dnn_lay=N_out_lab_cd
dnn_drop=0.0
dnn_use_laynorm_inp=False
dnn_use_batchnorm_inp=False
dnn_use_batchnorm=False
dnn_use_laynorm=False
dnn_act=softmax

arch_lr = 0.0002
arch_halving_factor=0.5
arch_improvement_threshold=0.001
arch_opt=rmsprop
opt_momentum=0.0
opt_alpha=0.95
opt_eps=1e-8
opt_centered=False
opt_weight_decay=0.0


[model]
model_proto=proto/model.proto
model:out_dnn1=compute(TRANSFORMER_AM,fmllr)
      out_dnn2=compute(MLP_layers,out_dnn1)
      loss_final=cost_nll(out_dnn2,lab_cd)
      err_final=cost_err(out_dnn2,lab_cd)


[forward]
forward_out=out_dnn2
normalize_posteriors=True
normalize_with_counts_from=lab_cd
save_out_file=False
require_decoding=True


[decoding]
decoding_script_folder=kaldi_decoding_scripts/
decoding_script=decode_dnn.sh
decoding_proto=proto/decoding.proto
min_active=200
max_active=7000
max_mem=50000000
beam=20.0
latbeam=12.0
acwt=0.10
max_arcs=-1
skip_scoring=false
scoring_script=/home/bin.yang02/software/kaldi-master/egs/librispeech/s5/local/score.sh
scoring_opts="--min-lmwt 4 --max-lmwt 23"
norm_vars=False
```

but when I only use simlpe mlp nn to train with the same feat/lab as input, the run_exp. py ran normally. 

Is there any configs missed？

**Comment by andi611** [comment]
Hi,

I haven't encountered this error before. The error msg doesn't tell much either.
However, I have a hunch to suspect this problem may be caused by downsampling and upsampling.
To verify this, can you try to save a randomly initialized aalbert model with `downsample_rate: 1`:
1. set `downsample_rate: 1` and `save_step: 1` in your .yaml
2. run the training script over 1 step then ctrl+c, this will save a randomly initialized model at training step 1
3. try to load this new model and see if the error still occurs.

**Please let me know if this is the case!**

FYI, we find that `downsample_rate: 1` is more suitable for the pytorch-kaldi DNN/HMM framework.
Using a downsample rate > 1 always yield worse results.

**Comment by ybNo1** [comment]
> Hi,
> 
> I haven't encountered this error before. The error msg doesn't tell much either.
> However, I have a hunch to suspect this problem may be caused by downsampling and upsampling.
> To verify this:
> 
> **Method 1** - can you try to save a randomly initialized aalbert model with `downsample_rate: 1`:
> 
> 1. set `downsample_rate: 1` and `save_step: 1` in your .yaml
> 2. run the training script over 1 step then ctrl+c, this will save a randomly initialized model at training step 1
> 3. try to load this new model and see if the error still occurs.
> 
> **Method 2** - in your `pytorch-kaldi/nn_transfromer.py`, print and verify that the input feature length and output representation length are the same.
> 
> 1. add `print(x.shape)` at both the beginning and end of the [forward function](https://github.com/andi611/Self-Supervised-Speech-Pretraining-and-Representation-Learning/blob/8ddb4fb17ae2a4942a6eac875b60c3a173516861/transformer/nn_transformer.py#L238) of your `pytorch-kaldi/nn_transfromer.py`.
> 2. execute the pytorch-kaldi `run_exp.py`
> 3. The input acoustic feature should have the shape of: (`time steps`, `batch size=12`, `dim=40`);
>    and the output representation should have the shape of (`time steps`, `batch size=12`, `dim=768`).
>    The `time steps` number should be the same.
> 
> **Please let me know if this is the case!**
> 
> FYI, we find that `downsample_rate: 1` is more suitable for the pytorch-kaldi DNN/HMM framework.
> Using a downsample rate > 1 always yield worse results.

thanks for replying, method One works for me!
Could you tell me what experiments you had done training ASR?About which structure of ASR you trained after pre-trained with aalbert/mockingjay,
I mean if aalbert acts the similar role with tdnn, then I should train ASR with simple fc layers and aalbert pretrain structure ,and ASR could aquire the same or better than kaldi tdnn ASR?

**Comment by ybNo1** [comment]
I've got answers from another issue, thanks for replying!I'll close this issue.

**Comment by andi611** [comment]
> thanks for replying, method One works for me!

I've tried setting downsample to 3 to match your setting, but I did not encounter any error.
Upsampling and downsampling worked fine, not sure why there is an error in your case.

Also, I found that the AALBERT config file is not set according to the original paper.
I've made some adjustments and updates in this [commit](https://github.com/andi611/Self-Supervised-Speech-Pretraining-and-Representation-Learning/commit/9cb533587e875456cc0ab4eda002da693bc58231).
Note that the previous settings are for our new work TERA, and not AALBERT.
I suggest you use the updated default config for future AALBERT training.

Happy training!
